File: frontend/src/types/intelligence.ts
export interface ScenarioProfile {
  predicted_yield: number
  lower_bound: number
  upper_bound: number
  spread: number
  uncertainty_pct: number
  risk_score: number
  risk_level: string
  is_anomaly: boolean
  anomaly_score: number
  features: Record<string, number>
}

export interface ChangedFeatureItem {
  feature_key: string
  feature_name: string
  baseline_value: number
  scenario_value: number
  absolute_change: number
  percent_change: number
}

export interface ScenarioSimulationDelta {
  yield_delta_kg_ha: number
  yield_percent_change: number
  risk_delta: number
  spread_delta_kg_ha: number
  direction: 'positive' | 'negative' | 'neutral'
  risk_direction: 'increased' | 'decreased' | 'unchanged'
}

export interface ScenarioSimulationRequest {
  year?: number
  state?: string
  state_code?: number
  district?: string
  baseline_rice_area?: number
  scenario_rice_area?: number
  scenario_total_cropped_area?: number
  scenario_rice_area_share?: number
  scenario_wheat_area?: number
  scenario_cotton_area?: number
  scenario_sugarcane_area?: number
  scenario_historical_yield_lag?: number
  scenario_rolling_yield?: number
}

export interface ScenarioSimulationResponse {
  state: string
  district: string
  year: number
  baseline: ScenarioProfile
  scenario: ScenarioProfile
  delta: ScenarioSimulationDelta
  changed_features: ChangedFeatureItem[]
  explanation: string
  warnings: string[]
  disclaimer: string
}

export interface CopilotQueryRequest {
  question: string
  context?: Record<string, any>
}

export interface CopilotEvidenceItem {
  source_name: string
  description: string
  records_count: number
  data_snippet?: any
}

export interface CopilotQueryResponse {
  question: string
  intent: string
  answer: string
  findings: string[]
  evidence: CopilotEvidenceItem[]
  tools_used: string[]
  records_analyzed: number
  model_outputs: any[]
  limitations: string[]
}

export interface ReportGenerateRequest {
  state?: string
  district?: string
  year?: number
  report_type?: string
}

export interface ReportGenerateResponse {
  report_id: string
  report_title: string
  report_type: string
  generated_at: string
  state?: string
  district?: string
  year?: number
  markdown_content: string
  summary_metrics: Record<string, any>
}

export interface RegionalSituationItem {
  state: string
  risk_level: string
  risk_score: number
  avg_yield: number
  volatility: number
  action_note: string
}

export interface ModelSignalItem {
  signal_name: string
  importance_pct: number
  description: string
}

export interface DecisionSupportResponse {
  kpis: {
    total_records: number
    anomalies_detected: number
    high_risk_states_count: number
    moderate_risk_states_count: number
    low_risk_states_count: number
    average_uncertainty_pct: number
    highest_risk_states: any[]
    recent_anomalies: any[]
  }
  regional_situation: RegionalSituationItem[]
  model_signals: ModelSignalItem[]
  recent_anomalies: any[]
  prediction_outlook: {
    national_mean_yield: number
    forecast_confidence: string
    active_pipeline: string
  }
  scenario_snapshot: {
    default_state: string
    simulated_intervention: string
    modeled_yield_shift: string
    risk_shift: string
  }
  ai_insight: string
  scientific_disclaimer: string
}

File: backend/schemas/intelligence.py
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, Field

# =========================================================================
# PART 1: SCENARIO SIMULATION SCHEMAS
# =========================================================================

class ScenarioSimulationRequest(BaseModel):
    year: int = Field(2017, description="Agricultural simulation year")
    state: Optional[str] = Field("Punjab", description="State Name")
    state_code: Optional[int] = Field(None, description="State Code")
    district: Optional[str] = Field("Ludhiana", description="District Name")
    # Base inputs
    baseline_rice_area: Optional[float] = Field(None, description="Optional baseline override ('000 ha)")
    # Scenario modifications
    scenario_rice_area: Optional[float] = Field(None, description="Modified rice area ('000 ha)")
    scenario_total_cropped_area: Optional[float] = Field(None, description="Modified total cropped land ('000 ha)")
    scenario_rice_area_share: Optional[float] = Field(None, description="Modified rice share (0.0 - 1.0)")
    scenario_wheat_area: Optional[float] = Field(None, description="Modified wheat area ('000 ha)")
    scenario_cotton_area: Optional[float] = Field(None, description="Modified cotton area ('000 ha)")
    scenario_sugarcane_area: Optional[float] = Field(None, description="Modified sugarcane area ('000 ha)")
    scenario_historical_yield_lag: Optional[float] = Field(None, description="Modified prior-year yield (kg/ha)")
    scenario_rolling_yield: Optional[float] = Field(None, description="Modified 3-yr rolling yield (kg/ha)")

class ScenarioProfile(BaseModel):
    predicted_yield: float
    lower_bound: float
    upper_bound: float
    spread: float
    uncertainty_pct: float
    risk_score: float
    risk_level: str
    is_anomaly: bool
    anomaly_score: float
    features: Dict[str, float]

class ChangedFeatureItem(BaseModel):
    feature_key: str
    feature_name: str
    baseline_value: float
    scenario_value: float
    absolute_change: float
    percent_change: float

class ScenarioSimulationDelta(BaseModel):
    yield_delta_kg_ha: float
    yield_percent_change: float
    risk_delta: float
    spread_delta_kg_ha: float
    direction: str
    risk_direction: str

class ScenarioSimulationResponse(BaseModel):
    state: str
    district: str
    year: int
    baseline: ScenarioProfile
    scenario: ScenarioProfile
    delta: ScenarioSimulationDelta
    changed_features: List[ChangedFeatureItem]
    explanation: str
    warnings: List[str]
    disclaimer: str

# =========================================================================
# PART 2: COPILOT & QUERY SCHEMAS
# =========================================================================

class CopilotQueryRequest(BaseModel):
    question: str = Field(..., min_length=2, max_length=1000, description="Natural language question")
    context: Optional[Dict[str, Any]] = Field(None, description="Optional conversation context")

class CopilotEvidenceItem(BaseModel):
    source_name: str
    description: str
    records_count: int
    data_snippet: Optional[Any] = None

class CopilotQueryResponse(BaseModel):
    question: str
    intent: str
    answer: str
    findings: List[str]
    evidence: List[CopilotEvidenceItem]
    tools_used: List[str]
    records_analyzed: int
    model_outputs: List[Dict[str, Any]]
    limitations: List[str]

# =========================================================================
# PART 3: AI INTELLIGENCE REPORT SCHEMAS
# =========================================================================

class ReportGenerateRequest(BaseModel):
    state: Optional[str] = Field("Punjab", description="Target State")
    district: Optional[str] = Field("Ludhiana", description="Target District")
    year: Optional[int] = Field(2017, description="Target Agricultural Year")
    report_type: str = Field("comprehensive", description="regional | state | district | model | risk | anomaly | comprehensive")

class ReportGenerateResponse(BaseModel):
    report_id: str
    report_title: str
    report_type: str
    generated_at: str
    state: Optional[str] = None
    district: Optional[str] = None
    year: Optional[int] = None
    markdown_content: str
    summary_metrics: Dict[str, Any]

# =========================================================================
# PART 4: DECISION SUPPORT SCHEMAS
# =========================================================================

class RegionalSituationItem(BaseModel):
    state: str
    risk_level: str
    risk_score: float
    avg_yield: float
    volatility: float
    action_note: str

class ModelSignalItem(BaseModel):
    signal_name: str
    importance_pct: float
    description: str

class DecisionSupportResponse(BaseModel):
    kpis: Dict[str, Any]
    regional_situation: List[RegionalSituationItem]
    model_signals: List[ModelSignalItem]
    recent_anomalies: List[Dict[str, Any]]
    prediction_outlook: Dict[str, Any]
    scenario_snapshot: Dict[str, Any]
    ai_insight: str
    scientific_disclaimer: str
